pkg/daemon/ceph/osd/kms/vault.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kms
+
+import (
+	"io/ioutil"
+	"strings"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/libopenstorage/secrets"
+	"github.com/libopenstorage/secrets/vault"
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// EtcVaultDir is vault config dir
+	EtcVaultDir = "/etc/vault"
+)
+
+var (
+	vaultMandatoryConnectionDetails = []string{api.EnvVaultAddress}
+	vaultTLSConnectionDetails       = []string{api.EnvVaultCACert, api.EnvVaultClientCert, api.EnvVaultClientKey}
+)
+
+/* VAULT API INTERNAL VALUES
+   const EnvVaultAddress = "VAULT_ADDR"
+   const EnvVaultAgentAddr = "VAULT_AGENT_ADDR"
+   const EnvVaultCACert = "VAULT_CACERT"
+   const EnvVaultCAPath = "VAULT_CAPATH"
+   const EnvVaultClientCert = "VAULT_CLIENT_CERT"
+   const EnvVaultClientKey = "VAULT_CLIENT_KEY"
+   const EnvVaultClientTimeout = "VAULT_CLIENT_TIMEOUT"
+   const EnvVaultSRVLookup = "VAULT_SRV_LOOKUP"
+   const EnvVaultSkipVerify = "VAULT_SKIP_VERIFY"
+   const EnvVaultNamespace = "VAULT_NAMESPACE"
+   const EnvVaultTLSServerName = "VAULT_TLS_SERVER_NAME"
+   const EnvVaultWrapTTL = "VAULT_WRAP_TTL"
+   const EnvVaultMaxRetries = "VAULT_MAX_RETRIES"
+   const EnvVaultToken = "VAULT_TOKEN"
+   const EnvVaultMFA = "VAULT_MFA"
+   const EnvRateLimit = "VAULT_RATE_LIMIT"
+*/
+
+// InitVault inits the secret store
+func InitVault(context *clusterd.Context, namespace string, config map[string]string) (secrets.Secrets, error) {
```

Comment:
A general comment... There is so much vault-specific code, is there any way we can generalize it? If each KMS provider requires this much customization, we cannot realistically support more than one. I was hoping the library would help abstract the provider, but it doesn't look like it helped in that aspect.